### DISCUSSION TITLE
`Condition<T>` and `ConditionedList<T>`

### DISCUSSION THREAD

**Discussion by ModernRonin** [sharpliner/sharpliner#166] [Ideas]
Hi again :-)

I just had the following scenario:

I wanted to create an extension method that fluently makes jobs go onto a specific VmImage pool. So first I wrote

```csharp
public static T OnLinux<T>(this T self) where T : JobBase =>
        self with { Pool = new Conditioned<Pool>(new HostedPool { VmImage = "ubuntu-latest" }) };
```

which works fine. But then I wanted to apply this to the whole pipeline and because there is no corresponding `Pool` property on pipeline definitions - which would be another issue, but, I think, a minor one -, I thought "okay, so let's create another extension method that I can use on `Jobs.`

Normally, if `Jobs` was a regular IEnumerable/List/whatever, I could just do
```csharp
public static ConditionedList<JobBase> OnLinux(this ConditionedList<JobBase> self) =>
      self.Select(e => e..Definition.OnLinux()).ToList();
```
or similar.
But that doesn't work because `Jobs` is a `ConditionedList<JobBase>` and `Conditioned<>.Definition` is `internal`.

Now I am aware that in this specific scenario, the simpler way to address my original use-case would be to expose a `Pool` property on `PipelineDefinition`s. But I think the scenario serves well to illustrate a more general point: the more easily users can extend the library in their local usage with extension methods, the less work there will be for library maintainers to cater to maybe very specific use-cases. 

I'm not sure what the best solution here is because I got not enough understanding yet what would break compatibility too much, so I'm starting a discussion here.

Somewhat related, more general: it would feel more natural for me if the collection types would be `ImmutableList`, `ImmutableDictionary` etc because that plays more naturally with the immutable nature of `record`s.

So if `Conditioned<T>` had some `public` way of creating a `{with ...}` copy and `ConditionedList<T>` was, say, derived from `ImmutableList<T>`, then I could write my extension method.

Sorry for this being just a brain-dump, I don't really have the time right now to look into this more deeply, but didn't want to forget it.

Cheers & "diky moc! :-)",
MR

**Comment by premun** [comment]
Hey,

Thanks for the feedback. You are totally right that extensibility would be very nice. I can see it myself with pipelines I am using Sharpliner for that each pipeline has its own distinct repeating features that could be solved with custom extensions. I am aware of this but there are several reasons where this is very problematic due to some internal Sharpliner reasons which is not quite obvious. I also understand that it is not obvious that since you get many benefits of C#, you get the impressions that you can do anything. However, there are many limitations and unless you don't know how Sharpliner works inside, it's hard to tell where is the line of what is and what is not possible.

The main reasons are:

1. `ConditionedList<T>` or even `Conditioned<T>` is actually not a flat list or a single item. It is a tree and a bit complicated tree too. This is due to the conditional expressions. Take for example the `Pool` property which is `Conditioned<Pool>`. Something like this is technically possible in the YAML:
    ```yaml
      pool:
        ${{ if eq(A, B) }}:
          name: pool-A
        ${{ else }}:
          name: pool-B
    ```
    This means that we need this to be a `Conditioned<T>`:
    ```csharp
    Pool = If.Equal("A", "B")
        .Pool(new HostedPool("pool-A"))
    .Else
        .Pool(new HostedPool("pool-B")),
    ```

    Since this can nest even more, it means that we actually need to store an expression tree. For `ConditionedList<T>` (which I think inherits from a `List<T>`) we have a list of trees.
    
    This means that your approach wouldn't work because you would need to search the whole tree and set the `VmImage` for all nodes. This already gets complicated and probably would have some other edge cases that would break that I cannot even think of.

2. It is quite expensive to add these shortcut methods and extensions such as `Pool()` as they need to be in several places and accept several types. Example: https://github.com/sharpliner/sharpliner/blob/main/src/Sharpliner/AzureDevOps/AzureDevOpsDefinition.cs These methods usually make sure we link the tree together well so that we can parse it. For instance when you are adding something into the `ConditionedList` and you specify a full if/else tree of things, the value that is actually being added into the list is the last leaf. So the `Add` method of the `ConditionedList` is overloaded and traverses the tree to the root and adds the root to the list instead.

3. I have to really make sure that people create the objects the right way so that they get linked into the tree. If I allowed people to create `Conditioned` or to touch `Definition` freely, they might break this mechanism and their definitions might disappear. So I have to take care about what I make public actually.

Hacks like these make it possible to implement the if/else grammar - and I was actually quite happy when I was able to even support that - but they also impose some non-obvious limitations that have some bad implications for the flexibility.

For your case, when you want to have the same pipeline twice with different pools, you could maybe utilize [Definition Libraries](https://github.com/sharpliner/sharpliner/blob/main/docs/AzureDevOps/DefinitionLibraries.md). You could create a template for your pipeline and make the pool a parameter which you would just set in the pipeline itself. You can then either make the whole pipeline a "library" or you could just have a job/step library which you can then reference in your pipelines using the `*Library()` methods. Libraries are basically a pattern when you want to generate different variants of a similar pipeline pieces.

Hopefully this will scratch your itch but I don't know your scenario in full.
